gateway/platforms/vertx3/vertx3/src/main/java/io/apiman/gateway/platforms/vertx3/engine/VertxPluginRegistry.java

```diff
@@ -72,30 +93,73 @@ public VertxPluginRegistry(Vertx vertx, VertxEngineConfig vxEngineConfig, Map<St
+
         int port = artifactUrl.getPort();
-        if (port == -1) {
-            port = 80;
+
+        //Configure http client options following artifact url
+        if (artifactUrl.getProtocol().equals("https")) {
+            //Enable SSL
+            clientOpts.setSsl(true);
+
+            //If port is not defined, set to https default port 443
+            if (port == -1) port = 443;
+
+            //If artifact host is localhost
+            if(artifactUrl.getHost().equals("localhost") ||artifactUrl.getHost().equals("127.0.0.1")){
+                //Reset proxy options (otherwise Vert.X try to use proxy for local connection)
+                clientOpts.setProxyOptions(null);
+            }else{
+                //Set SSL proxy options (if exists)
+                if(sslProxy!=null) clientOpts.setProxyOptions(sslProxy);
+            }
+
+        }else{
+            //Disable SSL
+            clientOpts.setSsl(false);
+
+            //If port is not defined, set to http default port 80
+            if (port == -1) port = 80;
+
+            //If artifact host is localhost
+            if(artifactUrl.getHost().equals("localhost") ||artifactUrl.getHost().equals("127.0.0.1")){
+                //Reset proxy options (otherwise Vert.X try to use proxy for local connection)
+                clientOpts.setProxyOptions(null);
+            }else{
+                //Set proxy options (if exists)
+                if(proxy!=null) clientOpts.setProxyOptions(proxy);
+            }
         }
 
+        //Create HTTP client with options
+        client = vertx.createHttpClient(clientOpts);
+
         final HttpClientRequest request = client.get(port, artifactUrl.getHost(), artifactUrl.getPath(),
                 (Handler<HttpClientResponse>) response -> {
 
-            response.exceptionHandler((Handler<Throwable>) error -> {
-                handler.handle(AsyncResultImpl.create(error, File.class));
-            });
-
-            // Body Handler
-            response.bodyHandler((Handler<Buffer>) buffer -> {
-                try {
-                    Files.write(pluginFile.toPath(), buffer.getBytes(), StandardOpenOption.APPEND,
-                            StandardOpenOption.CREATE, StandardOpenOption.WRITE);
-                    handler.handle(AsyncResultImpl.create(pluginFile));
-                } catch (IOException e) {
-                    handler.handle(AsyncResultImpl.create(e, File.class));
-                }
-            });
-        });
+                    response.exceptionHandler((Handler<Throwable>) error -> {
+                        handler.handle(AsyncResultImpl.create(error, File.class));
+                    });
+
+                    // Body Handler
+                    response.bodyHandler((Handler<Buffer>) buffer -> {
+                        try {
+                            // If status code is bad, do not handle the buffer.
+                            if(response.statusCode() != 200 ){
```

Comment:
What non-200 code are you receiving in the main response that still causes a body to be received here? i.e. why the special handling?

Is this because you need to handle 20x rather than just 200?

Reply:
During tests, I received 404 errors from Maven Central by querying unexisting artifact, example : https://repo1.maven.org/maven2/io/apiman/test/testPlugin/1.0.0.Final/testPlugin-1.0.0.Final.war

In this case, http status code is 404, and body content is (content-length : 554) :
```html
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
<!-- a padding to disable MSIE and Chrome friendly error page -->
```

With previous code, whatever status code, Vert.X gateway try anyway to extract plugin from the body. Il was reporting a wrong error "error in opening zip file".

By handling 200 code, we could be sure that the body really contain à valid plugin artifact.

Notate that the same control already exist in DefaultPluginRegistry :
https://github.com/apiman/apiman/blob/ddaf009dcb06fc6924616bfa63f725d090f62792/gateway/engine/core/src/main/java/io/apiman/gateway/engine/impl/DefaultPluginRegistry.java#L376-L387


DefaultPluginRegistry.java

Reply:
Okay, I see. We should probably handle non-20x as an error condition.